source/common/router/config_impl.cc

```diff
@@ -698,48 +692,66 @@ void RouteEntryImplBase::finalizeRequestHeaders(Http::RequestHeaderMap& headers,
-  } else {
-    // Most specific mutations (route-level) take precedence by being applied
-    // last: if a header is specified at all levels, the last one applied wins.
-    mergeTransforms(transforms,
-                    vhost_.globalRouteConfig().responseHeaderParser().getHeaderTransforms(
-                        stream_info, do_formatting));
-    mergeTransforms(transforms,
-                    vhost_.responseHeaderParser().getHeaderTransforms(stream_info, do_formatting));
-    mergeTransforms(transforms,
-                    response_headers_parser_->getHeaderTransforms(stream_info, do_formatting));
+  for (const HeaderParser* header_parser : getHeaderParsers(
+           /*get_request_header_parser=*/false,
+           /*specificity_ascend=*/vhost_.globalRouteConfig().mostSpecificHeaderMutationsWins())) {
+    // Later evaluated header parser wins.
+    mergeTransforms(transforms, header_parser->getHeaderTransforms(stream_info, do_formatting));
+  }
+  return transforms;
+}
+
+Http::HeaderTransforms
+RouteEntryImplBase::requestHeaderTransforms(const StreamInfo::StreamInfo& stream_info,
+                                            bool do_formatting) const {
+  Http::HeaderTransforms transforms;
+  for (const HeaderParser* header_parser : getHeaderParsers(
+           /*get_request_header_parser=*/true,
+           /*specificity_ascend=*/vhost_.globalRouteConfig().mostSpecificHeaderMutationsWins())) {
+    // Later evaluated header parser wins.
+    mergeTransforms(transforms, header_parser->getHeaderTransforms(stream_info, do_formatting));
   }
   return transforms;
 }
 
+std::vector<const HeaderParser*>
+RouteEntryImplBase::getHeaderParsers(bool get_request_header_parser,
+                                     bool specificity_ascend) const {
+  std::vector<const HeaderParser*> parsers;
+  const HeaderParser* global_route_config_header_parser =
+      get_request_header_parser ? &vhost_.globalRouteConfig().requestHeaderParser()
+                                : &vhost_.globalRouteConfig().responseHeaderParser();
+  const HeaderParser* vhost_header_parser =
+      get_request_header_parser ? &vhost_.requestHeaderParser() : &vhost_.responseHeaderParser();
+  const HeaderParser* route_header_parser =
+      get_request_header_parser ? request_headers_parser_.get() : response_headers_parser_.get();
+  if (specificity_ascend) {
+    // Appends user-specified request headers from least to most specific: global connection manager
+    // level headers, virtual host level headers and finally route-level headers.
+    parsers.push_back(global_route_config_header_parser);
+    parsers.push_back(vhost_header_parser);
+    parsers.push_back(route_header_parser);
+  } else {
+    // Appends user-specified request headers from most to least specific.
+    parsers.push_back(route_header_parser);
+    parsers.push_back(vhost_header_parser);
+    parsers.push_back(global_route_config_header_parser);
+  }
+  return parsers;
+}
+
```

Comment:
If I understand correctly, this function is to avoid duplication of code for request headers and response headers. Can we create some template code to make this code simple and more efficient?

Reply:
Yes, this function aims to avoid this function is to avoid duplication of code between request headers and response headers. 

Thanks for the advice on using template. My understanding with C++ template is that it is used to avoid duplicate between different data types with same/similar APIs. Our scenario here however is the reverse (i.e. request header parser and response header parser are of the same type but are accessed through different API). So I am not sure how template code can help in this case? (An example, if any, would be helpful).